Change CLI backend option to ValueEnum

aesteve-rh opened the pull request:
Change the CLI to receive backends as a positional
argument with values listed in a ValueEnum.

Current '--help' output:
```
A virtio-sound device using the vhost-user protocol.

Usage: vhost-user-sound --socket <SOCKET> --backend <BACKEND>

Options:
      --socket <SOCKET>  vhost-user Unix domain socket path
      --backend <BACKEND>  audio backend to be used [possible values:
      null, pipewire, alsa]
  -h, --help             Print help
  -V, --version          Print version
```

If a wrong backend is given, it give hints:
```
$ cargo run -- --socket /tmp/sound.sock --backend nul
error: invalid value 'nul' for '<BACKEND>'
  [possible values: null, pipewire, alsa]

  tip: a similar value exists: 'null'
```
Add a test to verify (minimally) the backend argument.
Last commit adds `rstest` crate to have multiple cases in a parametrized test.
Last commit outputs:
```
running 4 tests
test tests::test_cli_backend_arg::case_2_pipewire ... ok
test tests::test_cli_backend_arg::case_3_alsa ... ok
test tests::test_cli_backend_arg::case_1_null_backend ... ok
test tests::test_sound_config_setup ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```
But if the dependency is not desired, the commit can be omitted.

aesteve-rh committed:
sound: cli backend selector

Change the CLI backend option to receive values
listed in a ValueEnum.

Current '--help' output:
```
A virtio-sound device using the vhost-user protocol.

Usage: vhost-user-sound --socket <SOCKET> --backend <BACKEND>

Options:
      --socket <SOCKET>  vhost-user Unix domain socket path
      --backend <BACKEND>  audio backend to be used [possible values:
      null, pipewire, alsa]
  -h, --help             Print help
  -V, --version          Print version
```

If a wrong backend is given, it give hints:
```
$ cargo run -- --socket /tmp/sound.sock --backend nul
error: invalid value 'nul' for '<BACKEND>'
  [possible values: null, pipewire, alsa]

  tip: a similar value exists: 'null'
```

Signed-off-by: Albert Esteve <aesteve@redhat.com>

aesteve-rh committed:
sound: test cli backend argument

Signed-off-by: Albert Esteve <aesteve@redhat.com>

aesteve-rh committed:
sound: introduce rstest crate

Introduce 'rstest'[1] crate dependency to allow
run multiple parametrized tests (and fixtures).

[1] - https://docs.rs/rstest/latest/rstest/

Signed-off-by: Albert Esteve <aesteve@redhat.com>

bilelmoussaoui commented on crates/sound/src/lib.rs:
```diff
@@ -223,6 +223,10 @@ impl SoundConfig {
+
+    pub fn get_backend_name(&self) -> String {
+        String::from(&self.audio_backend_name)
+    }
```
the audio_backend should probably replaced to take BackendType & returned here instead of a string

aesteve-rh replied:
`get_backend_name` is a getter for the `SoundConfig.audio_backend_name`, which is a `String` declared as private.
I've just changed the type for `SoundArgs.backend` member.
Also, it is only used in a single test at the moment.
So I think the method returning a `String` is ok as is.

Now, another discussion would be if we want the backend stored in `SoundConfig` to be `BackendType`. I think it may be nice, completely dropping the `str`/`String` type even for the patterns here: https://github.com/virtio-sound/vhost-device/blob/virtio-sound/crates/sound/src/audio_backends.rs#L53
And matching `BackendType`s instead.

bilelmoussaoui replied:
Right, that is why I asked this. If an enum is introduced to define the backend to use, it should be used elsewhere as well.

aesteve-rh replied:
If I take this road, though, I would have to move the `BackendType` declaration to the `lib` module, including the clap `ValueEnum` attribute, which is a bit annoying as it is a CLI option that belongs to the binary.

Because otherwise, I cannot import stuff declared in main to other modules.

So I wonder if we really want to take this path.

bilelmoussaoui replied:
I think that is fine to use ValueEnum in lib.rs. At worst, you could also just implement the trait https://docs.rs/clap/latest/clap/trait.ValueEnum.html that the derive macro ValueEnum implements for you in main.rs. If you are lazy to do the implementation yourself, you can always use cargo expand the expand the macro and copy the generated code, clean it up a bit and done.

aesteve-rh replied:
I cannot implement an imported trait on an imported struct. But ok, I'll use the `ValueEnum` attribute in `lib`. Let's see how it looks.

bilelmoussaoui replied:
ooh indeed. Actually i think that is fine as clap is not explicitly set as a binary only dependency :)

aesteve-rh replied:
True :)

Done with the last push!

bilelmoussaoui commented on crates/sound/src/lib.rs:
```diff
@@ -90,6 +89,14 @@ impl From<stream::Error> for Error {
     }
 }
 
+#[derive(ValueEnum, Clone, Default, Debug, Eq, PartialEq)]
```
```suggestion
#[derive(ValueEnum, Clone, Copy, Default, Debug, Eq, PartialEq)]
```

bilelmoussaoui commented on crates/sound/src/lib.rs:
```diff
@@ -223,6 +230,10 @@ impl SoundConfig {
+
+    pub fn get_audio_backend(&self) -> &BackendType {
+        &self.audio_backend
+    }
```
```suggestion
    pub fn get_audio_backend(&self) -> BackendType {
        self.audio_backend
    }
```
It is a simple enum, with Copy derived you can drop the `&

aesteve-rh replied:
Or explicitely calling `self.audio_backend.clone()`. Not sure which option to favour in this case :shrug:

bilelmoussaoui replied:
It should be a copy type.

bilelmoussaoui reviewed (approved):
looks much better now, thanks a lot

stefano-garzarella commented on crates/sound/README.md:
````diff
@@ -30,7 +34,7 @@ generated with help2man target/debug/vhost-user-sound |mandoc
 
 ```shell
-host# vhost-user-sound --socket /tmp/snd.sock --backend null
+host# vhost-user-sound --socket /tmp/snd.sock null
````
I like the idea of using ValueEnum, but can we still have the `--backend` option?
I always found difficulty in positional arguments.

aesteve-rh replied:
Sure, will do :)

stefano-garzarella reviewed (commented):
Thanks for the PR!

I just noticed that `cargo build --no-default-features --features=null-backend` is now failing, can you check it?

aesteve-rh commented:
Fixed the build issue by tweaking the conditional compilation of the cfg attribute for the `match` block:
```rust
        #[cfg(not(all(
            feature = "null-backend",
            feature = "pw-backend",
            feature = "alsa-backend"
        )))]
        _ => Err(Error::AudioBackendNotSupported),
```
Does not look great, but I couldn't come with a cleaner correct option. Hopefully it is ok though :)

stefano-garzarella commented:
> Fixed the build issue by tweaking the conditional compilation of the cfg attribute for the `match` block:
> 
> ```rust
>         #[cfg(not(all(
>             feature = "null-backend",
>             feature = "pw-backend",
>             feature = "alsa-backend"
>         )))]
>         _ => Err(Error::AudioBackendNotSupported),
> ```
> 
> Does not look great, but I couldn't come with a cleaner correct option. Hopefully it is ok though :)

What about as was before this series, I mean applying this patch:

```patch
diff --git a/crates/sound/src/audio_backends.rs b/crates/sound/src/audio_backends.rs
index bd13c37..dea74f9 100644
--- a/crates/sound/src/audio_backends.rs
+++ b/crates/sound/src/audio_backends.rs
@@ -17,13 +17,7 @@ use self::alsa::AlsaBackend;
 use self::null::NullBackend;
 #[cfg(feature = "pw-backend")]
 use self::pipewire::PwBackend;
-use crate::{device::ControlMessage, stream::Stream, BackendType, Result};
-#[cfg(not(all(
-    feature = "null-backend",
-    feature = "pw-backend",
-    feature = "alsa-backend"
-)))]
-use crate::Error;
+use crate::{device::ControlMessage, stream::Stream, BackendType, Error, Result};
 
 pub trait AudioBackend {
     fn write(&self, stream_id: u32) -> Result<()>;
@@ -64,11 +58,6 @@ pub fn alloc_audio_backend(
         Pipewire => Ok(Box::new(PwBackend::new(streams))),
         #[cfg(feature = "alsa-backend")]
         Alsa => Ok(Box::new(AlsaBackend::new(streams))),
-        #[cfg(not(all(
-            feature = "null-backend",
-            feature = "pw-backend",
-            feature = "alsa-backend"
-        )))]
         _ => Err(Error::AudioBackendNotSupported),
     }
 }
```

stefano-garzarella commented:
> > Fixed the build issue by tweaking the conditional compilation of the cfg attribute for the `match` block:
> > ```rust
> >         #[cfg(not(all(
> >             feature = "null-backend",
> >             feature = "pw-backend",
> >             feature = "alsa-backend"
> >         )))]
> >         _ => Err(Error::AudioBackendNotSupported),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Does not look great, but I couldn't come with a cleaner correct option. Hopefully it is ok though :)
> 
> What about as was before this series, I mean applying this patch:
> 
> ```diff
> diff --git a/crates/sound/src/audio_backends.rs b/crates/sound/src/audio_backends.rs
> index bd13c37..dea74f9 100644
> --- a/crates/sound/src/audio_backends.rs
> +++ b/crates/sound/src/audio_backends.rs
> @@ -17,13 +17,7 @@ use self::alsa::AlsaBackend;
>  use self::null::NullBackend;
>  #[cfg(feature = "pw-backend")]
>  use self::pipewire::PwBackend;
> -use crate::{device::ControlMessage, stream::Stream, BackendType, Result};
> -#[cfg(not(all(
> -    feature = "null-backend",
> -    feature = "pw-backend",
> -    feature = "alsa-backend"
> -)))]
> -use crate::Error;
> +use crate::{device::ControlMessage, stream::Stream, BackendType, Error, Result};
>  
>  pub trait AudioBackend {
>      fn write(&self, stream_id: u32) -> Result<()>;
> @@ -64,11 +58,6 @@ pub fn alloc_audio_backend(
>          Pipewire => Ok(Box::new(PwBackend::new(streams))),
>          #[cfg(feature = "alsa-backend")]
>          Alsa => Ok(Box::new(AlsaBackend::new(streams))),
> -        #[cfg(not(all(
> -            feature = "null-backend",
> -            feature = "pw-backend",
> -            feature = "alsa-backend"
> -        )))]
>          _ => Err(Error::AudioBackendNotSupported),
>      }
>  }
> ```

Okay, in this case we will have `warning: unreachable pattern`, so I think your solution is the only one available, or we should suppress the `warn(unreachable_patterns)`

stefano-garzarella commented on crates/sound/src/audio_backends.rs:
```diff
@@ -46,17 +52,23 @@ pub trait AudioBackend {
 ) -> Result<Box<dyn AudioBackend + Send + Sync>> {
-    log::trace!("allocating audio backend {}", name);
-    match name.as_str() {
+    use BackendType::*;
```
I'd suggest avoiding this, since `cargo build --no-default-features ` is complaining about unused import.

What about this?
```patch
diff --git a/crates/sound/src/audio_backends.rs b/crates/sound/src/audio_backends.rs
index bd13c37..0d62591 100644
--- a/crates/sound/src/audio_backends.rs
+++ b/crates/sound/src/audio_backends.rs
@@ -55,15 +49,14 @@ pub fn alloc_audio_backend(
     backend: BackendType,
     streams: Arc<RwLock<Vec<Stream>>>,
 ) -> Result<Box<dyn AudioBackend + Send + Sync>> {
-    use BackendType::*;
     log::trace!("allocating audio backend {:?}", backend);
     match backend {
         #[cfg(feature = "null-backend")]
-        Null => Ok(Box::new(NullBackend::new(streams))),
+        BackendType::Null => Ok(Box::new(NullBackend::new(streams))),
         #[cfg(feature = "pw-backend")]
-        Pipewire => Ok(Box::new(PwBackend::new(streams))),
+        BackendType::Pipewire => Ok(Box::new(PwBackend::new(streams))),
         #[cfg(feature = "alsa-backend")]
-        Alsa => Ok(Box::new(AlsaBackend::new(streams))),
+        BackendType::Alsa => Ok(Box::new(AlsaBackend::new(streams))),
         #[cfg(not(all(
             feature = "null-backend",
             feature = "pw-backend",
```

aesteve-rh replied:
Yes, I could make the import optional too, but only three branches benefit from it, so it makes more sense to remove it as in your suggestion.

Anyway, does it make sense to compile without features? Maybe the Null backend should be always available.
In any case, it is out of the scope of this PR, so I will just follow your suggestion :)

stefano-garzarella replied:
Yes, maybe it doesn't make sense, and we should have the null backend always compiled. It's just that we have a feature to disable the null backend when not needed, and so I don't know how to make sure that there is always a compiled backend.

Maybe playing with Cargo.toml can be done, but I don't know if it's worth it ;-)

aesteve-rh replied:
Without features it also warns about unused variable with
```rust
streams: Arc<RwLock<Vec<Stream>>>,
```
Should I change it to `_streams` (in a separate commit)?

stefano-garzarella replied:
It seems pre-existing, so you don't have to, but if you want to do it that would be great ;-)

Also in this case, maybe the easiest way is to suppress the warning.

stefano-garzarella commented:
> Okay, in this case we will have `warning: unreachable pattern`, so I think your solution is the only one available, or we should suppress the `warn(unreachable_patterns)`

I think it might be better to suppress the warning for that line by putting a nice comment explaining why. Otherwise every time we add a backend we have to edit these lines.

WDYT?

aesteve-rh commented:
> I think it might be better to suppress the warning for that line by putting a nice comment explaining why. Otherwise every time we add a backend we have to edit these lines.
>
> WDYT?

I was tempted of doing that as I was trying to make it work. But I wanted to present the other option first.
I think suppressing the warning is best for this case, it is justified and will make it more maintainable. Let's go for that!

aesteve-rh committed:
sound: audio backend allow unused

When compiled with no features, the `streams`
parameter of the alloc_audio_backend() function
is not used. Mark it as allowed to avoid the
compilation warning.

Signed-off-by: Albert Esteve <aesteve@redhat.com>

aesteve-rh commented:
Another option now that I think of it, is to have the enum variants conditional in the declaration.

stefano-garzarella commented:
@aesteve-rh sorry, just noticed that `cargo fmt --check` suggests some changes that we may want to apply.

The rest LGTM ;-)

aesteve-rh commented:
> Another option now that I think of it, is to have the enum variants conditional in the declaration.

Nah, it complicates the `no-features` case. But it may be a nice change if we ever tweak the compiler to use at least the `Null` backend in any case (if it is possible).
I'll leave it as is.

stefano-garzarella commented:
> Another option now that I think of it, is to have the enum variants conditional in the declaration.

Do you mean in `pub enum BackendType` ?
It could make sense.

Should we also avoid the default?

aesteve-rh commented:
> Do you mean in pub enum BackendType ?
It could make sense.
>
>Should we also avoid the default?

Yes, the `BackendType`. And yes, we would have to avoid the default. That would've been fine, but
also it would leave the enum with no values with `no-features` case. And that makes some parts of
the code unreachable...
Kindof a mess to fix. But also proves that is not a good idea to compile with no features I guess :)

Let's keep it in the back of our heads for the future.

stefano-garzarella commented:
> > Do you mean in pub enum BackendType ?
> > It could make sense.
> > Should we also avoid the default?
> 
> Yes, the `BackendType`. And yes, we would have to avoid the default. That would've been fine, but also it would leave the enum with no values with `no-features` case. And that makes some parts of the code unreachable... Kindof a mess to fix. But also proves that is not a good idea to compile with no features I guess :)

I'm start thinking the same :-)
So maybe we should always compile the null backend.
Disable it will save just few bytes, nothing more since it doesn't have any dependency.

> 
> Let's keep it in the back of our heads for the future.

Sure, feel free to open an issue here for tracking it.

stefano-garzarella commented:
> > Another option now that I think of it, is to have the enum variants conditional in the declaration.
> 
> Nah, it complicates the `no-features` case. But it may be a nice change if we ever tweak the compiler to use at least the `Null` backend in any case (if it is possible). I'll leave it as is.

However, I think it makes sense, partly because now we print all as possible values, even if not enabled.

```bash
$ cargo build --no-default-features --features=null-backend
$ target/debug/vhost-user-sound --socket /tmp/sock --backend gstreamer
error: invalid value 'gstreamer' for '--backend <BACKEND>'
  [possible values: null, pipewire, alsa]

For more information, try '--help'.
```

But I agree that we can do it later by removing the feature to disable null-backend to simplify the code.

aesteve-rh commented:
> However, I think it makes sense, partly because now we print all as possible values, even if not enabled.

Right, that is a good point.

> But I agree that we can do it later by removing the feature to disable null-backend to simplify the code.

I don't mind handling this myself in a follow-up PR :)
